sdk/iot/Azure.Iot.Hub.Service/src/Authentication/SasTokenProviderWithSharedAccessKey.cs

```diff
@@ -0,0 +1,68 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+
+namespace Azure.Iot.Hub.Service.Authentication
+{
+    /// <summary>
+    /// Implementation of a shared access signature provider with token caching and refresh.
+    /// </summary>
+    internal class SasTokenProviderWithSharedAccessKey : ISasTokenProvider
+    {
+        private static readonly TimeSpan s_defaultTimeToLive = TimeSpan.FromMinutes(30);
+        private readonly object _lock = new object();
+
+        private readonly string _hostname;
+        private readonly string _sharedAccessPolicy;
+        private readonly string _sharedAccessKey;
+        private readonly TimeSpan _timeToLive;
+
+        private string _cachedSasToken;
+        private DateTimeOffset _tokenExpiryTime;
+
+        // Private constructor to prevent accidental initialzation.
+        private SasTokenProviderWithSharedAccessKey()
+        {
+        }
+
+        public SasTokenProviderWithSharedAccessKey(string hostname, string sharedAccessPolicy, string sharedAccessKey, TimeSpan? timeToLive = null)
+        {
```

Comment:
Do you want to allow negative timespan values? I don't see a check anywhere

Reply:
Good point, I'll add a check for that. 
I actually don't have a way for the user to set the sas token TimeToLive right now; maybe it should be customizable via `IotHubServiceClientOptions`?

Reply:
yep I can see users wanting to provide that as input.

Reply:
Just fyi @prmathur-microsoft  - This is internal and users should not be calling it. We are using it to get SAS token from the connection string provided by users.

Reply:
Yes Sindhu. But data is flowing from CS which is input from customer.

Reply:
I have a very basic question :) Do connection strings have the timeToLive on them or do we calculate it somehow? I will go and investigate one for IotHub as I've not paid a lot of attention to the values in a connection string earlier.

Reply:
@vinagesh no

Reply:
Tracked here: https://msazure.visualstudio.com/One/_workitems/edit/7462988